concretizer: refactor version, compiler version and platform rules

alalazo opened the pull request:
This adds an explicit rule to ensure that concrete versions from an abstract spec are always concretizable.

It also removes `versions_weights` in favor of `version/3` that contains a package, a version and a weights. Using the same function avoids by design to infer the weight from a declaration and the actual version from another.

Require that all selected versions are declared either in `package.py`, `packages.yaml` or in an abstract_spec.

Cardinality constraints on compiler versions and platforms are reworked as integrity constraints, to avoid looping on 2 variables in the grounder.

tgamblin commented:
Do we want to de-orthogonalize this?  It makes sense to me that the weights are specified separately from the fact that a package has a particular version.  If this ends up being a major performance improvement, I'd definitely reconsider but it seems beneficial to me to have difference concepts expressed separately in the logic program.  e.g, if we wanted to change how the versions are chosen and not just use weights but something else, we could easily separate that out if the rules are separate.  With this change I am not so sure.  How big a win is this?

alalazo committed:
concretizer: add a test on package version preferences

alalazo committed:
concretizer: rework version rules

This adds an explicit rule to ensure that concrete versions
from an abstract spec are always concretizable.

It also removes 'versions_weights' in favor of 'version/3'
that contains a package a version and a weight. Using the
same function avoids by design to infer the weight from a
declaration and the actual version from another.

Require that all selected versions are declared, either in
package.py, packages.yaml or abstract_spec.
